File: backend/internal/models/patient_checkup.go
package models

import "time"

type PatientCheckupMedicine struct {
	MedicineID   string  `json:"medicine_id"`
	MedicineName string  `json:"medicine_name"`
	Quantity     int     `json:"quantity"`
	Dosage       string  `json:"dosage"`
	Frequency    string  `json:"frequency"`
	DurationDays int     `json:"duration_days"`
	Notes        *string `json:"notes,omitempty"`
}

type PatientCheckup struct {
	BaseUUID

	PatientID string  `gorm:"type:uuid;not null;index" json:"patient_id"`
	Patient   Patient `gorm:"foreignKey:PatientID" json:"patient"`

	VisitDate      time.Time `gorm:"not null;index" json:"visit_date"`
	Status         string    `gorm:"type:varchar(20);not null;default:'scheduled';index" json:"status"` // scheduled, completed, cancelled
	ChiefComplaint string    `gorm:"type:text;not null" json:"chief_complaint"`
	Symptoms       []string  `gorm:"type:jsonb;serializer:json;not null" json:"symptoms"`

	Diagnosis        *string                  `gorm:"type:text" json:"diagnosis,omitempty"`
	TemperatureC     *float64                 `gorm:"type:decimal(5,2)" json:"temperature_c,omitempty"`
	BloodPressure    *string                  `gorm:"type:varchar(20)" json:"blood_pressure,omitempty"`
	HeartRate        *int                     `gorm:"check:heart_rate >= 0" json:"heart_rate,omitempty"`
	RespiratoryRate  *int                     `gorm:"check:respiratory_rate >= 0" json:"respiratory_rate,omitempty"`
	OxygenSaturation *int                     `gorm:"check:oxygen_saturation >= 0 AND oxygen_saturation <= 100" json:"oxygen_saturation,omitempty"`
	HeightCm         *float64                 `gorm:"type:decimal(6,2);check:height_cm >= 0" json:"height_cm,omitempty"`
	WeightKg         *float64                 `gorm:"type:decimal(6,2);check:weight_kg >= 0" json:"weight_kg,omitempty"`
	Medicines        []PatientCheckupMedicine `gorm:"type:jsonb;serializer:json" json:"medicines,omitempty"`
	TreatmentPlan    *string                  `gorm:"type:text" json:"treatment_plan,omitempty"`
	Notes            *string                  `gorm:"type:text" json:"notes,omitempty"`
	DoctorName       *string                  `gorm:"type:varchar(255)" json:"doctor_name,omitempty"`
	FollowUpDate     *time.Time               `gorm:"type:date" json:"follow_up_date,omitempty"`

	CreatedAt time.Time  `gorm:"autoCreateTime" json:"created_at"`
	UpdatedAt time.Time  `gorm:"autoUpdateTime" json:"updated_at"`
	DeletedAt *time.Time `gorm:"index" json:"deleted_at,omitempty"`
}

func (PatientCheckup) TableName() string {
	return "patient_checkups"
}

File: frontend/src/schemas/patient-checkup.ts
import { z } from "zod";

export const patientCheckupMedicineSchema = z.object({
  medicine_id: z.string().min(1, "Medicine is required"),
  medicine_name: z.string().min(1, "Medicine name is required"),
  quantity: z.coerce
    .number<number | string>()
    .min(1, "Quantity must be at least 1"),
  dosage: z.string().min(1, "Dosage is required"),
  frequency: z.string().min(1, "Frequency is required"),
  duration_days: z.coerce
    .number<number | string>()
    .min(1, "Duration must be at least 1 day"),
  notes: z.string().optional(),
});

export const createPatientCheckupSchema = z.object({
  patient_id: z.string().min(1, "Patient is required"),
  visit_date: z.string().min(1, "Visit date is required"),
  chief_complaint: z.string().min(1, "Chief complaint is required"),
  symptoms: z.string().min(1, "Symptoms are required"),
  status: z.enum(["scheduled", "completed", "cancelled"]).optional(),
  diagnosis: z.string().optional(),
  temperature_c: z.coerce.number<number | string>().min(30).max(45).optional(),
  blood_pressure: z.string().optional(),
  heart_rate: z.coerce.number<number | string>().min(0).optional(),
  respiratory_rate: z.coerce.number<number | string>().min(0).optional(),
  oxygen_saturation: z.coerce
    .number<number | string>()
    .min(0)
    .max(100)
    .optional(),
  height_cm: z.coerce.number<number | string>().min(0).optional(),
  weight_kg: z.coerce.number<number | string>().min(0).optional(),
  medicines: z.array(patientCheckupMedicineSchema).optional(),
  treatment_plan: z.string().optional(),
  doctor_name: z.string().optional(),
  follow_up_date: z.string().optional(),
  notes: z.string().optional(),
  patient_allergies: z.string().optional(),
  patient_blood_type: z
    .enum(["A+", "A-", "B+", "B-", "AB+", "AB-", "O+", "O-"])
    .optional()
    .or(z.literal("")),
});

export const updatePatientCheckupSchema = z.object({
  visit_date: z.string().min(1, "Visit date is required"),
  chief_complaint: z.string().min(1, "Chief complaint is required"),
  symptoms: z.string().min(1, "Symptoms are required"),
  status: z.enum(["scheduled", "completed", "cancelled"]),
  diagnosis: z.string().optional(),
  temperature_c: z.coerce.number<number | string>().min(30).max(45).optional(),
  blood_pressure: z.string().optional(),
  heart_rate: z.coerce.number<number | string>().min(0).optional(),
  respiratory_rate: z.coerce.number<number | string>().min(0).optional(),
  oxygen_saturation: z.coerce
    .number<number | string>()
    .min(0)
    .max(100)
    .optional(),
  height_cm: z.coerce.number<number | string>().min(0).optional(),
  weight_kg: z.coerce.number<number | string>().min(0).optional(),
  medicines: z.array(patientCheckupMedicineSchema).optional(),
  treatment_plan: z.string().optional(),
  doctor_name: z.string().optional(),
  follow_up_date: z.string().optional(),
  notes: z.string().optional(),
  patient_allergies: z.string().optional(),
  patient_blood_type: z
    .enum(["A+", "A-", "B+", "B-", "AB+", "AB-", "O+", "O-"])
    .optional()
    .or(z.literal("")),
});

export type CreatePatientCheckupFormInput = z.input<
  typeof createPatientCheckupSchema
>;
export type CreatePatientCheckupFormData = z.output<
  typeof createPatientCheckupSchema
>;
export type UpdatePatientCheckupFormInput = z.input<
  typeof updatePatientCheckupSchema
>;
export type UpdatePatientCheckupFormData = z.output<
  typeof updatePatientCheckupSchema
>;
export type PatientCheckupMedicineFormData = z.infer<
  typeof patientCheckupMedicineSchema
>;

export const patientCheckupStatusOptions = [
  { value: "scheduled", label: "Scheduled" },
  { value: "completed", label: "Completed" },
  { value: "cancelled", label: "Cancelled" },
] as const;
